File: src/moveit_visual_tools_demo.cpp
// Copyright 2015 University of Colorado, Boulder
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

/* Author: Dave Coleman
   Desc:   Demo implementation of moveit_visual_tools
           To use, add a Rviz Marker Display subscribed to topic /moveit_visual_tools
*/

// ROS
#include <rclcpp/rclcpp.hpp>
#include <ament_index_cpp/get_package_share_directory.hpp>

// For visualizing things in rviz
#include <moveit_visual_tools/moveit_visual_tools.h>

// MoveIt
#include <moveit/planning_scene_monitor/planning_scene_monitor.h>

// C++
#include <string>

// Boost
#include <boost/lexical_cast.hpp>

using namespace std::literals;

static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_visual_tools_demo");

namespace rvt = rviz_visual_tools;

namespace moveit_visual_tools
{
static const std::string PLANNING_GROUP_NAME = "arm";  // RRBot Specific
static const std::string THIS_PACKAGE = "moveit_visual_tools";

class VisualToolsDemo
{
public:
  /**
   * \brief Constructor
   */
  VisualToolsDemo(const rclcpp::Node::SharedPtr& node) : node_(node)
  {
    visual_tools_ = std::make_shared<moveit_visual_tools::MoveItVisualTools>(node_, "world", "/moveit_visual_tools");
    visual_tools_->loadPlanningSceneMonitor();
    visual_tools_->loadMarkerPub(true);
    visual_tools_->loadRobotStatePub("display_robot_state");
    visual_tools_->setManualSceneUpdating();

    robot_state_ = visual_tools_->getSharedRobotState();
    jmg_ = robot_state_->getJointModelGroup(PLANNING_GROUP_NAME);
    // Allow time to publish messages
    rclcpp::spin_some(node);
    rclcpp::sleep_for(1000ms);

    // Clear collision objects and markers
    visual_tools_->deleteAllMarkers();
    visual_tools_->removeAllCollisionObjects();
    visual_tools_->triggerPlanningSceneUpdate();
    rclcpp::sleep_for(1000ms);

    // Show message
    Eigen::Isometry3d text_pose = Eigen::Isometry3d::Identity();
    text_pose.translation().z() = 4;
    visual_tools_->publishText(text_pose, "MoveIt Visual Tools", rvt::WHITE, rvt::XLARGE, /*static_id*/ false);

    runRobotStateTests();

    runDeskTest();

    runCollisionObjectTests();
  }

  void publishLabelHelper(const Eigen::Isometry3d& pose, const std::string& label)
  {
    Eigen::Isometry3d pose_copy = pose;
    pose_copy.translation().x() -= 0.2;
    visual_tools_->publishText(pose_copy, label, rvt::WHITE, rvt::XLARGE, false);
    visual_tools_->trigger();
  }

  void runRobotStateTests()
  {
    // Show 5 random robot states
    RCLCPP_INFO(LOGGER, "Showing 5 random robot states");
    for (std::size_t i = 0; i < 5; ++i)
    {
      robot_state_->setToRandomPositions(jmg_);
      visual_tools_->publishRobotState(robot_state_, rvt::DEFAULT);
      rclcpp::sleep_for(1000ms);
    }

    // Show 5 robot state in different colors
    RCLCPP_INFO(LOGGER, "Showing 5 random robot states in different colors");
    for (std::size_t i = 0; i < 5; ++i)
    {
      robot_state_->setToRandomPositions(jmg_);
      visual_tools_->publishRobotState(robot_state_, visual_tools_->getRandColor());
      rclcpp::sleep_for(1000ms);
    }

    // Hide the robot
    RCLCPP_INFO(LOGGER, "Hiding the robot");
    visual_tools_->hideRobot();
    rclcpp::sleep_for(1000ms);

    // Show the robot
    RCLCPP_INFO(LOGGER, "Showing the robot");
    visual_tools_->publishRobotState(robot_state_, rvt::DEFAULT);
    rclcpp::sleep_for(1000ms);
  }

  void runDeskTest()
  {
    double common_angle = M_PI * 1.1;
    double x_offset = -3.0;

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Moving the robot");
    Eigen::Isometry3d robot_pose = Eigen::Isometry3d::Identity();
    robot_pose = robot_pose * Eigen::AngleAxisd(common_angle, Eigen::Vector3d::UnitZ());
    robot_pose.translation().x() = x_offset;
    visual_tools_->applyVirtualJointTransform(*robot_state_, robot_pose);
    visual_tools_->publishRobotState(robot_state_, rvt::DEFAULT);
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Floor");
    visual_tools_->publishCollisionFloor(-0.5, "Floor", rvt::GREY);
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Wall");
    double wall_x = x_offset - 1.0;
    double wall_y = -1.0;
    double wall_width = 6.0;
    double wall_height = 4;
    visual_tools_->publishCollisionWall(wall_x, wall_y, common_angle, wall_width, wall_height, "Wall", rvt::GREEN);
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Table");
    double table_x = x_offset + 1.0;
    double table_y = 0;
    double table_z = 0;
    double table_width = 3;
    double table_height = 1;
    double table_depth = 1;
    visual_tools_->publishCollisionTable(table_x, table_y, table_z, common_angle, table_width, table_height,
                                         table_depth, "Table", rvt::BLUE);
    rclcpp::sleep_for(1000ms);

    // Send ROS messages
    visual_tools_->triggerPlanningSceneUpdate();

    // Show 5 random robot states
    RCLCPP_INFO(LOGGER, "Showing 5 random robot states");
    for (std::size_t i = 0; i < 5; ++i)
    {
      robot_state_->setToRandomPositions(jmg_);
      visual_tools_->publishRobotState(robot_state_, rvt::DEFAULT);
      rclcpp::sleep_for(1000ms);
    }
  }

  void runCollisionObjectTests()
  {
    // Create pose
    Eigen::Isometry3d pose1 = Eigen::Isometry3d::Identity();
    Eigen::Isometry3d pose2 = Eigen::Isometry3d::Identity();

    double space_between_rows = 0.2;
    double y = 0;
    double step;

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Mesh");
    // TODO: Catch exception
    std::string file_path = "file://" + ament_index_cpp::get_package_share_directory(THIS_PACKAGE);
    if (file_path == "file://")
      RCLCPP_FATAL_STREAM(LOGGER, "Unable to get " << THIS_PACKAGE << " package path ");
    file_path.append("/resources/demo_mesh.stl");
    step = 0.4;
    for (double i = 0; i <= 1.0; i += 0.1)
    {
      visual_tools_->publishCollisionMesh(visual_tools_->convertPose(pose1),
                                          "Mesh_" + boost::lexical_cast<std::string>(i), file_path,
                                          visual_tools_->getRandColor());
      if (!i)
        publishLabelHelper(pose1, "Mesh");
      pose1.translation().x() += step;
    }
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Block");
    pose1 = Eigen::Isometry3d::Identity();
    y += space_between_rows * 2.0;
    pose1.translation().y() = y;
    for (double i = 0; i <= 1.0; i += 0.1)
    {
      double size = 0.1 * (i + 0.5);
      visual_tools_->publishCollisionBlock(visual_tools_->convertPose(pose1),
                                           "Block_" + boost::lexical_cast<std::string>(i), size,
                                           visual_tools_->getRandColor());

      if (!i)
        publishLabelHelper(pose1, "Block");
      pose1.translation().x() += step;
    }
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Rectanglular Cuboid");
    double cuboid_min_size = 0.05;
    double cuboid_max_size = 0.2;
    pose1 = Eigen::Isometry3d::Identity();
    pose2 = Eigen::Isometry3d::Identity();
    y += space_between_rows * 2.0;
    pose1.translation().y() = y;
    pose2.translation().y() = y;
    for (double i = 0; i <= 1.0; i += 0.1)
    {
      pose2 = pose1;
      pose2.translation().x() += i * cuboid_max_size + cuboid_min_size;
      pose2.translation().y() += (i * cuboid_max_size + cuboid_min_size) / 2.0;
      pose2.translation().z() += i * cuboid_max_size + cuboid_min_size;
      visual_tools_->publishCollisionCuboid(pose1.translation(), pose2.translation(),
                                            "Rectangle_" + boost::lexical_cast<std::string>(i),
                                            visual_tools_->getRandColor());
      if (!i)
        publishLabelHelper(pose1, "Cuboid");
      pose1.translation().x() += step;
    }
    rclcpp::sleep_for(1000ms);

    // --------------------------------------------------------------------
    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Cylinder");
    double cylinder_min_size = 0.01;
    double cylinder_max_size = 0.3;
    pose1 = Eigen::Isometry3d::Identity();
    pose2 = Eigen::Isometry3d::Identity();
    y += space_between_rows * 2.0;
    pose1.translation().y() = y;
    pose2.translation().y() = y;
    for (double i = 0; i <= 1.0; i += 0.1)
    {
      double radius = 0.1;
      pose2 = pose1;
      pose2.translation().x() += i * cylinder_max_size + cylinder_min_size;
      // pose2.translation().y() += i * cylinder_max_size + cylinder_min_size;
      // pose2.translation().z() += i * cylinder_max_size + cylinder_min_size;
      visual_tools_->publishCollisionCylinder(pose1.translation(), pose2.translation(),
                                              "Cylinder_" + boost::lexical_cast<std::string>(i), radius,
                                              visual_tools_->getRandColor());

      if (!i)
        publishLabelHelper(pose1, "Cylinder");
      pose1.translation().x() += step;

      pose1 = pose1 * Eigen::AngleAxisd(step * 2 * M_PI, Eigen::Vector3d::UnitZ());
    }

    // TODO(davetcoleman): publishCollisionGraph
    // TODO(davetcoleman): publishContactPoint
    // TODO(davetcoleman): trajectory stuff
    // TODO(davetcoleman): gripper stuff

    // Send ROS messages
    visual_tools_->triggerPlanningSceneUpdate();
    rclcpp::sleep_for(1000ms);
  }

private:
  // A shared node handle
  rclcpp::Node::SharedPtr node_;

  // For visualizing things in rviz
  moveit_visual_tools::MoveItVisualToolsPtr visual_tools_;

  // MoveIt Components
  planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor_;

  const moveit::core::JointModelGroup* jmg_;

  moveit::core::RobotStatePtr robot_state_;
};  // end class

}  // namespace moveit_visual_tools

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto node = rclcpp::Node::make_shared("moveit_visual_tools_demo", node_options);
  RCLCPP_INFO_STREAM(LOGGER, "Visual Tools Demo");

  moveit_visual_tools::VisualToolsDemo demo(node);

  RCLCPP_INFO_STREAM(LOGGER, "Shutting down.");
  rclcpp::shutdown();

  return 0;
}

File: python/examples/moveit_visual_tools_demo.py
#!/usr/bin/env python3

import moveit_visual_tools as mvt
import rviz_visual_tools as rvt
import rclcpp
import sys
import time
from rclpy import logging
import numpy as np
from ament_index_python.packages import get_package_share_directory
from copy import deepcopy

logger = logging.get_logger("moveit_visual_tools_demo")

from geometry_msgs.msg import Pose, Point, Quaternion

rclcpp.init()

PLANNING_GROUP_NAME = "arm"

node = rclcpp.Node("moveit_visual_tools_demo")
visual_tools = mvt.MoveItVisualTools(node, "world", "/moveit_visual_tools")
visual_tools.load_planning_scene_monitor()
visual_tools.load_marker_publisher(True)
visual_tools.load_robot_state_publisher("display_robot_state")
visual_tools.set_manual_scene_updating()

robot_state = visual_tools.get_shared_robot_state()
robot_model = visual_tools.get_robot_model()
jmg = robot_model.get_joint_model_group(PLANNING_GROUP_NAME)

visual_tools.delete_all_markers()
visual_tools.remove_all_collision_objects()
visual_tools.trigger_planning_scene_update()

visual_tools.publish_text(
    Pose(position=Point(x=0.0, y=0.0, z=4.0)),
    "MoveIt-Visual-Tools",
    scale=rvt.Scales.XXLARGE,
)
visual_tools.trigger()

logger.info("Showing 5 random robot states")
for _ in range(5):
    robot_state.set_to_random_positions(jmg)
    visual_tools.publish_robot_state(robot_state)
    time.sleep(1.0)

logger.info("Showing 5 random robot states in different colors")
for _ in range(5):
    robot_state.set_to_random_positions(jmg)
    visual_tools.publish_robot_state(robot_state, color=visual_tools.get_random_color())
    time.sleep(1.0)

logger.info("Hiding robot state")
visual_tools.hide_robot()
time.sleep(1.0)

logger.info("Showing robot state")
visual_tools.publish_robot_state(robot_state, color=rvt.Colors.DEFAULT)
time.sleep(1.0)

logger.info("Publishing Collision Floor")
visual_tools.publish_collision_floor(-0.5, "Floor", color=rvt.Colors.GREY)
visual_tools.trigger_planning_scene_update()
time.sleep(1.0)

logger.info("Publishing Collision Wall")
visual_tools.publish_collision_wall(
    -4.0, -1.0, 0.0, np.pi * 1.1, 6.0, 4.0, "Wall", color=rvt.Colors.GREY
)
visual_tools.trigger_planning_scene_update()
time.sleep(1.0)


logger.info("Publishing Collision Mesh")
mesh_pose = Pose()
for i in range(5):
    visual_tools.publish_collision_mesh(
        mesh_pose,
        f"Mesh_{i}",
        f"file://{get_package_share_directory('moveit_visual_tools')}/resources/demo_mesh.stl",
        color=visual_tools.get_random_color(),
    )
    mesh_pose.position.x += 0.4
    visual_tools.trigger_planning_scene_update()
time.sleep(1.0)

logger.info("Publishing Collision Block")
block_pose = Pose(position=Point(y=1.0))
for i in np.linspace(0.0, 1.0, 10):
    visual_tools.publish_collision_block(
        block_pose,
        f"Block_{i}",
        color=visual_tools.get_random_color(),
    )
    block_pose.position.x += 0.4
    visual_tools.trigger_planning_scene_update()
time.sleep(1.0)

logger.info("Publishing Collision Rectanglular Cuboid")
cuboid_min_size = 0.05
cuboid_max_size = 0.2
cuboid_point1 = Point(y=2.0)
for i in np.linspace(0.0, 1.0, 10):
    cuboid_point2 = deepcopy(cuboid_point1)
    cuboid_point2.x += i * cuboid_max_size + cuboid_min_size
    cuboid_point2.y += (i * cuboid_max_size + cuboid_min_size) / 2.0
    cuboid_point2.z += i * cuboid_max_size + cuboid_min_size
    visual_tools.publish_collision_cuboid(
        cuboid_point1, cuboid_point2, f"Cuboid_{i}", visual_tools.get_random_color()
    )
    cuboid_point1.x += 0.4
    visual_tools.trigger_planning_scene_update()
time.sleep(1.0)

logger.info("Publishing Collision Cylinder")
cylinder_min_size = 0.01
cylinder_max_size = 0.3
cylinder_pose = Pose(position=Point(y=3.0))
for i in np.linspace(0.0, 1.0, 10):
    visual_tools.publish_collision_cylinder(
        cylinder_pose,
        height=i * cylinder_max_size + cylinder_min_size,
        radius=i * cylinder_max_size + cylinder_min_size,
        object_name=f"Cylinder_{i}",
        color=visual_tools.get_random_color(),
    )
    cylinder_pose.position.x += 0.1 + i
    visual_tools.trigger_planning_scene_update()
time.sleep(1.0)

rclcpp.shutdown()
